### PR TITLE
Improve plugin load and unload ordering

### PR DESCRIPTION
**Pull request by RhysB**
# 📌 Pull Request

## Description

Improve plugin load, enable, and disable ordering by introducing a plugin load order planner.

Behaviour:
- Plans the plugin load order utilizing the plugin.yml information.
- Makes the load order deterministic to prevent OS/Java types from effecting load order.
- Priority hard dependencies over soft dependencies where possible.
- Disable plugins in the reverse order.

## Related Issues / Discussions

Fixes issues discussed in the RetroMC Discord with Zavdav and RobertWesner

- Fixes #
- Related to #

## Motivation & Context

Fixes several bugs regarding plugin loading, enabling, and disabling.

## Type of Change

Please tick all that apply:

- [x] 🐛 Bug fix (non-breaking change that fixes an issue)
- [ ] ⚡ Performance improvement
- [ ] 🧩 New feature (non-breaking)
- [ ] 🔧 Refactor / cleanup (no functional changes)
- [ ] 🔥 Crash / exploit fix
- [ ] 📚 Documentation update
- [ ] ❗ Breaking change (may affect plugins or server behavior)

## Testing Performed

- [x] Compiled successfully with `mvn clean package`
- [x] Tested on a Beta 1.7.3 server
- [x] Existing functionality verified
- [x] Edge cases considered

**Test details:**
```text
- Tested on the RetroMC development server with approx 70 plugins.
- PR resolves load issues we have previously had with some of our plugins.
```

## Compatibility Considerations

- [ ] No known compatibility impact
- [x] Potential plugin impact (described below)
- [ ] Network / protocol (Netcode) behavior changed
- [ ] Configuration change required

## Compatibility Notes

This PR shouldn't break compatibility of plugins with correct plugin.yml files.

## Checklist

Please confirm the following:
- [x] No unnecessary formatting or whitespace-only changes
- [x] Changes are compatible with Minecraft Beta 1.7.3
- [x] No dependencies have been added without discussion with the RetroMC team

### CODE REVIEW

File: src/main/java/org/bukkit/plugin/SimplePluginManager.java [hunk 1]
```diff
@@ -147,51 +149,24 @@ public Plugin[] loadPlugins(File directory) {
...
-            if (finalPass) {
-                break;
-            } else if (allFailed) {
-                finalPass = true;
+            } catch (UnknownDependencyException ex) {
+                server.getLogger().log(Level.SEVERE, "Could not load '" + plannedPlugin.file.getPath() + "' in folder '" + directory.getPath() + "': " + ex.getMessage(), ex);
+            } catch (InvalidPluginException ex) {
+                server.getLogger().log(Level.SEVERE, "Could not load '" + plannedPlugin.file.getPath() + "' in folder '" + directory.getPath() + "': ", ex.getCause());
+            } catch (InvalidDescriptionException ex) {
+                server.getLogger().log(Level.SEVERE, "Could not load '" + plannedPlugin.file.getPath() + "' in folder '" + directory.getPath() + "': " + ex.getMessage(), ex);
```
Comment (RobertWesner) [inline src/main/java/org/bukkit/plugin/SimplePluginManager.java]: - multi-catch would be beneficial here, as all catches are (almost¹) exactly the same
- ¹ InvalidDescriptionException is missing the message but has the colon and using `ex.getCause()` instead of `ex`
- in my opinion you don't need to attach the previous exception message, the cause exists for that purpose, the information is practically included twice.

```suggestion
            } catch (UnknownDependencyException | InvalidPluginException | InvalidDescriptionException ex) {
                server.getLogger().log(Level.SEVERE, "Could not load '" + plannedPlugin.file.getPath() + "' in folder '" + directory.getPath() + "'.", ex);
```

File: src/main/java/org/bukkit/plugin/SimplePluginManager.java [hunk 2]
```diff
@@ -318,15 +293,29 @@ public void enablePlugin(final Plugin plugin) {
 
             try {
                 plugin.getPluginLoader().enablePlugin(plugin);
+                // Record successful enables so shutdown can run in strict reverse dependency order.
+                if (plugin.isEnabled() && !enableOrder.contains(plugin)) {
```
Comment (RobertWesner) [inline src/main/java/org/bukkit/plugin/SimplePluginManager.java]: In what scenario could the second condition `!enableOrder.contains(plugin)` be relevant?
It could only ever reach here when its not already enabled.
Sounds like its a defensive check that will never actually happen ^^

File: src/main/java/org/bukkit/craftbukkit/CraftServer.java [hunk 3]
```diff
@@ -163,6 +164,72 @@ public void disablePlugins() {
         pluginManager.disablePlugins();
     }
 
+    private boolean shouldAttemptEnable(Plugin plugin, PluginLoadOrder type) {
+        // Startup plugins are eligible during both passes. Postworld plugins only become eligible later.
+        //TODO Reevaluate if this should be allowed
+        return plugin.getDescription().getLoad().ordinal() <= type.ordinal();
+    }
+
+    private boolean enablePlugin(Plugin plugin, PluginLoadOrder type, Set<String> enabling) {
```
Comment (RobertWesner) [inline src/main/java/org/bukkit/craftbukkit/CraftServer.java:176]: I don't fully understand why this complexity is needed.

Is it just to ensure startup plugins get loaded first and remaining plugins get loaded afterwards?

Why not just .filter() the plugin load order beforehand into those that run on startup and those who run post world load?

How is a a startup plugin depending on postworld handled?

File: src/main/java/com/legacyminecraft/poseidon/PluginLoadPlanner.java [hunk 4]
```diff
@@ -0,0 +1,303 @@
...
+                }
+
+                stream = jar.getInputStream(entry);
+                return new PluginDescriptionFile(stream);
+            } catch (IOException ex) {
+                throw new InvalidPluginException(ex);
+            } catch (YAMLException ex) {
+                throw new InvalidPluginException(ex);
+            } finally {
+                if (stream != null) {
```
Comment (RobertWesner) [inline src/main/java/com/legacyminecraft/poseidon/PluginLoadPlanner.java]: you could use try-with-resources since java7 on both stream and jar, which in my opinion is much cleaner than checking for being null in finally

File: src/main/java/com/legacyminecraft/poseidon/PluginLoadPlanner.java [hunk 5]
```diff
@@ -0,0 +1,303 @@
...
+
+        return plan;
+    }
+
+    private PluginCandidate createCandidate(File file) {
+        PluginDescriptionFile description;
+
+        try {
+            description = getPluginDescription(file);
+        } catch (InvalidPluginException ex) {
```
Comment (RobertWesner) [inline src/main/java/com/legacyminecraft/poseidon/PluginLoadPlanner.java]: this also could be a multi catch, and InvalidPluginException is missing the message (which imo can be omitted from both) and is using ex.getCause instead of ex

File: src/main/java/com/legacyminecraft/poseidon/PluginLoadPlanner.java [hunk 6]
```diff
@@ -0,0 +1,303 @@
...
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.*;
+import java.util.jar.JarEntry;
+import java.util.jar.JarFile;
+import java.util.logging.Level;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+public final class PluginLoadPlanner {
```
Comment (RobertWesner) [inline src/main/java/com/legacyminecraft/poseidon/PluginLoadPlanner.java:19]: Just here to confirm it works, even if the ordering differs from mine, the dependencies are still maintained:

```
A ABC B C D E F G H

B depends on D
C depends on D
D depends on A
H depends on B
ABC depends on A, B, C
```

Image

File: src/main/java/org/bukkit/plugin/SimplePluginManager.java [hunk 7]
```diff
@@ -39,6 +40,7 @@ public final class SimplePluginManager implements PluginManager {
     private final Map<Boolean, Set<Permission>> defaultPerms = new LinkedHashMap<Boolean, Set<Permission>>();
     private final Map<String, Map<Permissible, Boolean>> permSubs = new HashMap<String, Map<Permissible, Boolean>>();
     private final Map<Boolean, Map<Permissible, Boolean>> defSubs = new HashMap<Boolean, Map<Permissible, Boolean>>();
+    private final List<Plugin> enableOrder = new ArrayList<Plugin>();
```
Comment (RobertWesner) [inline src/main/java/org/bukkit/plugin/SimplePluginManager.java]: Why do we have the generic types on all `new`? It can be inferred by compiler (and IDE). Just stylistic choice?

File: src/main/java/org/bukkit/craftbukkit/CraftServer.java [hunk 8]
```diff
@@ -147,8 +147,9 @@ public void enablePlugins(PluginLoadOrder type) {
         Plugin[] plugins = pluginManager.getPlugins();
 
         for (Plugin plugin : plugins) {
-            if ((!plugin.isEnabled()) && (plugin.getDescription().getLoad() == type)) {
-                loadPlugin(plugin);
+            // Re-evaluate every disabled plugin on each phase so deferred dependencies can come alive later.
+            if ((!plugin.isEnabled()) && shouldAttemptEnable(plugin, type)) {
```
Comment (RobertWesner) [inline src/main/java/org/bukkit/craftbukkit/CraftServer.java]:
```suggestion
            if (!plugin.isEnabled() && shouldAttemptEnable(plugin, type)) {
```

**Review by RobertWesner — CHANGES_REQUESTED** [review]
Works, but the code has some minor edges that we might want to change.
